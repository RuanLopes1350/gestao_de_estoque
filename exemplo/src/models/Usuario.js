// models/Usuario.js
import mongoose from "mongoose";
import mongoosePaginate from 'mongoose-paginate-v2';

/**
 * Necessario manter os imports para Grupo e Unidade, mesmo sem uso aqui na classe pois o populate do mongoose não funciona se não houver a referencia
 */
import Grupo from './Grupo.js';
import Unidade from './Unidade.js';

class Usuario {
  constructor() {
    const usuarioSchema = new mongoose.Schema(
      {
        nome: { type: String, index: true },
        email: { type: String, unique: true},
        senha: { type: String, select: false },
        link_foto: { type: String },
        ativo: { type: Boolean, default: false },
        tokenUnico: { type: String, select: false }, // Token único para validação de email, recuperação de senha e autenticação
        refreshtoken: { type: String, select: false }, // Refresh token para geração de access token de autenticação longa duração 7 dias para invalidação
        accesstoken: { type: String, select: false }, // Refresh token para  autenticação curta longa 15 minutos para invalidação
        // Referências para Unidades
        codigo_recupera_senha: { type: String, select: false, unique: true }, // Código de recuperação de senha, usado para validar a recuperação de senha do usuário
        exp_codigo_recupera_senha: { type: Date, select: false }, // Data de expiração do código de recuperação de senha, usado para validar a recuperação de senha do usuário

        unidades: [
          {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'unidades' // nome da coleção dentro do banco
          }
        ],

        // Referências para Grupos
        grupos: [
          {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'grupos',
          }
        ],

        /**
         * Permissões personalizadas para cada rota, a rota e o domínio 
         * devem ser únicos e conrrespondentes a rota e domínio do sistema em /rotas
         */
        permissoes: [
          {
            rota: { type: String, index: true, required: true }, // usuários / grupos / unidades / rotas
            dominio: { type: String }, // http://localhost:3000
            ativo: { type: Boolean, default: false },  // false
            buscar: { type: Boolean, default: false },    // false
            enviar: { type: Boolean, default: false },   // false
            substituir: { type: Boolean, default: false },    // false
            modificar: { type: Boolean, default: false },  // false
            excluir: { type: Boolean, default: false }, // false
          }
        ],
      },
      {
        timestamps: true,
        versionKey: false
      }
    );

    // Validação personalizada para garantir que rota + dominio sejam únicos dentro do usuário isso tbm vai ser feito pelo ZOD
    usuarioSchema.pre('save', function (next) {
      const permissoes = this.permissoes;
      const combinacoes = permissoes.map(p => `${p.rota}_${p.dominio}`);
      const setCombinacoes = new Set(combinacoes);

      if (combinacoes.length !== setCombinacoes.size) {
        return next(new Error('Permissões duplicadas encontradas: rota + domínio devem ser únicos dentro de cada usuário.'));
      }

      next();
    });

    usuarioSchema.plugin(mongoosePaginate);

    this.model = mongoose.model('usuarios', usuarioSchema);
  }
}

export default new Usuario().model;
