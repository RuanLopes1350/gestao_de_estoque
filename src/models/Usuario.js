import mongoose from 'mongoose';
import mongoosePaginate from 'mongoose-paginate-v2';

export class Usuario {
    constructor() {
        const usuarioSchema = mongoose.Schema (
            {
                nome_usuario: {type: String, index: true, required: true},
                matricula: {type: String, index: true, required: true, unique: true},
                senha: {type: String, required: true, select: false},
                email: {type: String, required: true, unique: true},
                cargo: {type: String, required: true},
                status: {type: Boolean, default: true},
                acessToken: {type: String, required: true},
                refreshToken: {type: String, required: true},
                data_cadastro: {type: Date, default: Date.now},
                data_atualizacao: {type: Date, default: Date.now},
                data_ultimo_login: {type: Date, default: Date.now},
            },
            {
                versionKey: false,
                timestamps: {createdAt: 'data_cadastro', updatedAt: 'data_atualizacao'},
            }
        )
        usuarioSchema.plugin(mongoosePaginate);
        this.model = mongoose.model('usuarios', usuarioSchema);
    }
}

export default new Usuario().model;