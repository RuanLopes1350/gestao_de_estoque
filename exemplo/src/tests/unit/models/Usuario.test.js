// src/tests/unit/models/Usuario.test.js
import mongoose from 'mongoose';
import Usuario from '../../../../src/models/Usuario.js';
import { MongoMemoryServer } from 'mongodb-memory-server';

// src/tests/unit/models/Usuario.test.js

let mongoServer;

// Configuração antes de todos os testes
beforeAll(async () => {
    mongoServer = await MongoMemoryServer.create();
    const uri = mongoServer.getUri();

    await mongoose.connect(uri, {
        // Opções de conexão não são necessárias no Mongoose 6+
    });
});

// Limpeza após todos os testes
afterAll(async () => {
    await mongoose.disconnect();
    await mongoServer.stop();
});

// Limpeza após cada teste para garantir isolamento
afterEach(async () => {
    jest.clearAllMocks();
    await Usuario.deleteMany({});
});

describe('Modelo de Usuário', () => {
    it('deve criar um usuário com dados válidos', async () => {
        const userData = {
            nome: 'Test User',
            email: 'test@example.com',
            senha: 'password123',
            link_foto: 'http://example.com/photo.jpg',
            ativo: true,
            permissoes: [
                {
                    rota: 'usuarios',
                    dominio: 'http://localhost:3000',
                    ativo: true,
                    buscar: true,
                    enviar: true,
                    substituir: true,
                    modificar: true,
                    excluir: true
                }
            ]
        };

        const user = new Usuario(userData);
        await user.save();

        const savedUser = await Usuario.findById(user._id).select('-senha'); // Busca o usuário sem o campo senha

        expect(savedUser.nome).toBe(userData.nome);
        expect(savedUser.email).toBe(userData.email);
        expect(savedUser.senha).toBeUndefined(); // Verificação para confirmar que senha não está presente
        expect(savedUser.link_foto).toBe(userData.link_foto);
        expect(savedUser.ativo).toBe(userData.ativo);
        expect(savedUser.permissoes.length).toBe(1);
        expect(savedUser.permissoes[0].rota).toBe(userData.permissoes[0].rota);
    });

    it('não deve criar um usuário com email duplicado', async () => {
        const userData = {
            nome: 'Test User',
            email: 'duplicate@example.com',
            senha: 'password123'
        };

        const user1 = new Usuario(userData);
        await user1.save();

        const user2 = new Usuario(userData);
        await expect(user2.save()).rejects.toThrowErrorMatchingSnapshot();
    });

    it('não deve criar um usuário com rota e domínio duplicados em permissões', async () => {
        const userData = {
            nome: 'Test User',
            email: 'test2@example.com',
            senha: 'password123',
            permissoes: [
                {
                    rota: 'usuarios',
                    dominio: 'http://localhost:3000',
                    ativo: true,
                    buscar: true,
                    enviar: true,
                    substituir: true,
                    modificar: true,
                    excluir: true
                },
                {
                    rota: 'usuarios',
                    dominio: 'http://localhost:3000',
                    ativo: true,
                    buscar: true,
                    enviar: true,
                    substituir: true,
                    modificar: true,
                    excluir: true
                }
            ]
        };

        const user = new Usuario(userData);
        await expect(user.save()).rejects.toThrowError('Permissões duplicadas encontradas: rota + domínio devem ser únicos dentro de cada usuário.');
    });

    it('deve criar um usuário com múltiplas permissões válidas', async () => {
        const userData = {
            nome: 'Test User',
            email: 'test3@example.com',
            senha: 'password123',
            permissoes: [
                {
                    rota: 'usuarios',
                    dominio: 'http://localhost:3000',
                    ativo: true,
                    buscar: true,
                    enviar: true,
                    substituir: true,
                    modificar: true,
                    excluir: true
                },
                {
                    rota: 'grupos',
                    dominio: 'http://localhost:3000',
                    ativo: true,
                    buscar: true,
                    enviar: true,
                    substituir: true,
                    modificar: true,
                    excluir: true
                }
            ]
        };

        const user = new Usuario(userData);
        await user.save();

        const savedUser = await Usuario.findById(user._id).select('-senha');

        expect(savedUser.permissoes.length).toBe(2);
        expect(savedUser.permissoes[0].rota).toBe(userData.permissoes[0].rota);
        expect(savedUser.permissoes[1].rota).toBe(userData.permissoes[1].rota);
    });

    it('deve criar um usuário com unidades e grupos válidos', async () => {
        const unidadeId = new mongoose.Types.ObjectId();
        const grupoId = new mongoose.Types.ObjectId();

        const userData = {
            nome: 'Test User',
            email: 'test4@example.com',
            senha: 'password123',
            unidades: [unidadeId],
            grupos: [grupoId]
        };

        const user = new Usuario(userData);
        await user.save();

        const savedUser = await Usuario.findById(user._id).select('-senha');

        expect(savedUser.unidades.length).toBe(1);
        expect(savedUser.unidades[0].toString()).toBe(unidadeId.toString());
        expect(savedUser.grupos.length).toBe(1);
        expect(savedUser.grupos[0].toString()).toBe(grupoId.toString());
    });
});

