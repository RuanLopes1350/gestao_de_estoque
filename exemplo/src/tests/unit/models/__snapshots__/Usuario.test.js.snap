// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modelo de Usuário não deve criar um usuário com email duplicado 1`] = `"E11000 duplicate key error collection: test.usuarios index: email_1 dup key: { email: "duplicate@example.com" }"`;
